runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,450 @@
+ * Elements whose transitive closure inner reference count matches the actual reference count are ones
+ * belonging to the garbage cycles and thus can be discarded.
+ * We ignore elements reachable from objects having external references (i.e. inner rc != real rc).
+ * If during computations of the aggregated RC there were modifications in the reference counts of
+ * elements of the atomic rootset:
+ *   - if it is being increased, then someone already got an external reference to this element, thus we may not
+ *     end up matching the inner reference count anyway
+ *   - if it is being decreased and object become garbage, it will be collected next time
+ * If transitive closure of the atomic rootset mutates, it could only happen via changing the atomics references,
+ * as all elements of this closure are frozen.
+ * To handle such mutations we keep collector flag, which is cleared before analysis and set on every
+ * atomic reference value update. If flag's value changes - collector restarts its analysis.
+ * There are not so much of complications in this algorithm due to the delayed reference counting as if there's a
+ * stack reference to the shared object - it's reflected in the reference counter (see rememberNewContainer()).
+ * We release objects found by the collector on a rendezvouz callback, but not on the main thread,
+ * to keep UI responsive, as taking GC lock can take time, sometimes.
+ */
+namespace {
+
+class Locker {
+  pthread_mutex_t* lock_;
+
+ public:
+  Locker(pthread_mutex_t* alock): lock_(alock) {
+    pthread_mutex_lock(lock_);
+  }
+
+  ~Locker() {
+    pthread_mutex_unlock(lock_);
+  }
+};
+
+template <typename func>
+inline void traverseObjectFields(ObjHeader* obj, func process) {
+  RuntimeAssert(obj != nullptr, "Must be non null");
+  const TypeInfo* typeInfo = obj->type_info();
+  if (typeInfo != theArrayTypeInfo) {
+    for (int index = 0; index < typeInfo->objOffsetsCount_; index++) {
+      ObjHeader** location = reinterpret_cast<ObjHeader**>(
+          reinterpret_cast<uintptr_t>(obj) + typeInfo->objOffsets_[index]);
+      process(location);
+    }
+  } else {
+    ArrayHeader* array = obj->array();
+    for (int index = 0; index < array->count_; index++) {
+      process(ArrayAddressOfElementAt(array, index));
+    }
+  }
+}
+
+inline bool isAtomicReference(ObjHeader* obj) {
+  return (obj->type_info()->flags_ & TF_LEAK_DETECTOR_CANDIDATE) != 0;
+}
+
+#define CHECK_CALL(call, message) RuntimeCheck((call) == 0, message)
+
+class CyclicCollector {
+  pthread_mutex_t lock_;
+  pthread_cond_t cond_;
+  pthread_t gcThread_;
+
+  int currentAliveWorkers_;
+  int gcRunning_;
+  int mutatedAtomics_;
+  int pendingRelease_;
+  bool shallRunCollector_;
+  bool terminateCollector_;
+  int32_t currentTick_;
+  int32_t lastTick_;
+  int64_t lastTimestampUs_;
+  void* mainWorker_;
+  KStdUnorderedMap<ObjHeader*, int> sideRefCounts_;
```

Comment:
Side ref counts should be per-container.

Reply:
No, as container can be changed during freezing.

Reply:
Leak reproducer:
```kotlin
import kotlin.native.concurrent.*

fun array(size: Int) = Array<Any?>(size, { null })

fun createGarbage() {
    val ref = AtomicReference<Any?>(null)
    val obj1 = array(2)
    val obj2 = array(1)
    val obj3 = array(2)

    obj1[0] = obj2
    obj1[1] = obj3

    obj2[0] = obj3

    obj3[0] = obj2
    obj3[1] = ref

    ref.value = obj1.freeze()
}

fun main() {
    createGarbage()
}
```